File: westcoast/src/main/java/frc/commands/targeting/StayOnTarget.java
package frc.commands.targeting;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.wpilibj2.command.PIDCommand;
import frc.subsystems.DriveTrainSubsystem;
import org.jetbrains.annotations.NotNull;

import static frc.utils.Maths.clamp;

public final class StayOnTarget extends PIDCommand {
    @NotNull
    private final DriveTrainSubsystem drive;

    public StayOnTarget(
            @NotNull PIDController controller,
            @NotNull DriveTrainSubsystem drive
    ) {
        super(
                controller,
                drive.vision::getXOffset,
                0.0,
                (value) -> drive.arcade(0.0, clamp(value, 0.4), false),
                drive
        );
        this.drive = drive;
    }

    @Override
    public void end(boolean interrupted) {
        drive.stop();
    }
}

File: westcoast/src/main/kotlin/frc/commands/targeting/StayOnTarget.kt
package frc.commands.targeting

import edu.wpi.first.math.controller.PIDController
import edu.wpi.first.wpilibj2.command.PIDCommand
import frc.subsystems.DriveTrainSubsystem

class StayOnTarget(
    controller: PIDController,
    private val drive: DriveTrainSubsystem
) : PIDCommand(
    /* controller = */ controller,
    /* measurementSource = */ { drive.vision.xOffset },
    /* setpoint = */ 0.0,
    /* useOutput = */ { drive.arcade(0.0, it.coerceIn(-0.4, 0.4), false) },
    /* ...requirements = */ drive
) {
    override fun end(interrupted: Boolean) {
        drive.stop()
    }
}